Blending enhancements

hlecleme opened the pull request:
This PR proposes some enhancements on the blending operations of the modules. I'm grouping the changes together to reduce the number of `dt_develop_blend_params_t` structure changes.

It contains the following changes:
1. invert mask blurring and feathering to reduce halos
2. make a more rational UI for reverse blending modes (may fix #7464)
3. refactoring: move all “legacy” `dt_develop_blend_params_t` structures to the `dt_develop_blend_legacy_params` function
4. add blend fulcrum to RGB channel blend modes

**Invert mask blurring and feathering to reduce halos**

This is a suggestion from @aurelienpierre to reduce the halos generated by the blurring of the mask, when blurring is activated to “fill in” holes (e.g. caused by parametric masking). The choice whether to blur before feather or feather before blur is done through the “feathering guide” parameter:

![image](https://user-images.githubusercontent.com/50139228/107873908-23e14080-6eb6-11eb-99d3-25ad32ddeadd.png)

Based on some testing I've found that blurring and feathering are actually fighting each other and it is not easy to have a properly delimited mask that is smooth inside. Nevertheless, I've made the new order (feather after blur) the default as in most cases it is better to avoid the halos.

Note: it is perhaps possible to be more fine-grained in the operations, but I haven't yet found a way to provide this “configurability” without making the interface too complex.

**Make a more rational UI for reverse blending modes**

I finally decided to go for the “generic” implementation of reverse blending modes (i.e. blending the input image on top of the output image). This means that it is available for all blending modes (within all blending “color spaces”). It is available through the icon next to the blend mode parameter:

![image](https://user-images.githubusercontent.com/50139228/107874184-090fcb80-6eb8-11eb-8f28-c209b7896f6d.png)

There are two performance impacts:
1. It requires more memory: a whole additional image buffer is needed (like in the OpenCL path) to avoid aliasing one of the input buffers with the output buffer. Removing the `restrict` qualifier could make it possible to optimize this in the path for “normal” blending modes, but I'm not sure it is worth it.
2. It requires an additional copy of the whole (output) image (also like the OpenCL path). This could be improved in a future “pixel pipe” redesign (by considering the blending as a separate operation with two input images and one independent output image, that could be cached separately)

I'm not sure yet about the balance between that additional functionality and the performance impacts. If needed I could only implement a few of the blending modes as “reverse” blending modes in the old way, but I would like to avoid the need to add one or two of those at every release.

Feel free to try this out and give some feedback :)

dterrahe commented on data/kernels/blendop.cl:
```diff
@@ -21,8 +21,8 @@
-  DEVELOP_BLEND_DISABLED = 0x00,
-  DEVELOP_BLEND_NORMAL = 0x01, /* deprecated as it did clamping */
+  /* DEVELOP_BLEND_DISABLED = 0x00, same as the new normal */
+  /* DEVELOP_BLEND_NORMAL = 0x01, deprecated as it did clamping */
```
Why not define the values here and use them in _blend_legacy_blend_mode rather than having magic numbers there? If you want to make sure they are not used accidentally anywhere else, add _DEPRECATED to the name in both locations.

hlecleme replied:
Ok for me, done.

hlecleme commented:
@TurboGit could you have a quick look to see if this change is acceptable?
Thanks in advance!

TurboGit commented:
Sure looking at this.

![image](https://user-images.githubusercontent.com/467069/116906247-4b87af80-ac40-11eb-8734-6bbd4c186e98.png)

You have added an icon (next to multiply) to reverse the blending mode, but there is no tooltip. Can one be added? I suppose we need quite some explanations for this to be clear.

I have not looked at the code yet, but I agree with the 4 points above. The blend fulcrum is not well defined and so the tooltip "adjust the fulcrum used by some blending operations" is not very informative. I guess users is just invited to move the slider and see what is happening! Why not...

The blend GUI is certainly reaching a point where it is far too encumbered and we will probably need to be reorganised after release 3.6.

As we are on the blending section there is something I have just noticed and I don't really understand.

When selecting "drawn mask" we have a way to change the polarity of the masks (icon next to the combo with masks) and an "invert mask" combo after the shape selection.

![image](https://user-images.githubusercontent.com/467069/116907020-50009800-ac41-11eb-9b47-af50a6fabd1d.png)

What is the difference ? And do you understand why this "invert mask" combo is only available with "drawn mask" ? That is, this control is not on the "parametric mask" nor on the "draw + parametric" modes ! Something I may have missed....

hlecleme commented:
I'm not sure I can add a tooltip only for the “quad” that is painted next to the combo. But I could change the whole tooltip to:

    choose blending mode and the blending order between
    the input and the output of the module,
    by default the output will be blended on top of the input,
    when active it will be reversed (input on top of the output)

Is there a better (more concise?) way to explain it?

It is not really easy to explain the “blend fulcrum”. It can also be considered as a weight on one of the operands of the blending operation (e.g. `input * (fulcrum * output)` or `input + (fulcrum * output)`). But I'm not sure it will be more comprehensible to people not very familiar with the maths behind the blending operations... I'm open to suggestions!

For the polarity of the drawn masks, the mask polarity button and the invert mask combo are two different things but they have the same effect when used the "drawn mask" mode. The first one is an inversion applied after the computation of the mask using the shapes. The second one is used in the parametric mask computation (which is also used in the “drawn mask” mode) and inverts the result after performing the parametric mask computation (in “drawn mask” mode this would be only an inversion).

The combo “invert mask” is actually the “same” as the “combine masks” combo in the other two modes, this can be seen by performing the following steps: use “parametric mask” mode and set combine masks to “exclusive”, switch to “drawn mask” and change the invert mask combo from “off” to “on”, then when switching back to “parametric mask” mode the combine masks combo will be set to “exclusive & inverted”.

TurboGit commented:
> but they have the same effect when used the "drawn mask" mode.

Right, that was my point. So why not remove it ? This seems just confusing no ?

TurboGit commented:
> choose blending mode and the blending order between
> the input and the output of the module,
> by default the output will be blended on top of the input,
> when active it will be reversed (input on top of the output)

I propose to change last line to:
> order can be reversed by clicking on the icon (input on top of the output)

aurelienpierre commented on src/develop/blend.h:
```diff
@@ -25,7 +25,7 @@
 #include "gui/color_picker_proxy.h"
 
-#define DEVELOP_BLEND_VERSION (10)
+#define DEVELOP_BLEND_VERSION (11)
```
that's going to conflict with #8673

TurboGit replied:
Yep, you'll need to rebase your work on top of this when merged.

aurelienpierre replied:
No, I mean I will lose edits. Rebasing is easy.

TurboGit replied:
Oh I see, you are already using it ! You may have to create a temp legacy_params() to upgrade your edits then.

hlecleme commented:
The last commit adds a separate button for the blend order (instead of combining it with the blend modes combo). This has the advantage of allowing a separate tooltip message and being a little bit more consistent with the rest of the blending GUI:

![image](https://user-images.githubusercontent.com/50139228/116964747-87585e80-acac-11eb-8b65-c53c031483d7.png)

> but they have the same effect when used the "drawn mask" mode.
>> Right, that was my point. So why not remove it ? This seems just confusing no ?

They have the same effect when used individually. When inverting the shapes polarity and setting “on” to invert mask, the effect is a double inversion = the same as if the shapes and the mask were not inverted.

Now, if we remove the “invert mask” combo, depending on whether we set the “combine masks” combo to “exclusive” or “exclusive & inverted” when using the “drawn & parametric mask” mode, the display of the polarity for the shapes (in “drawn mask” mode) will mean something different. If we remove and *reset* the “invert mask” combo, then switching back and forth between modes (“drawn & parametric mask” and “drawn mask”) will not keep the mask the same (it could generate an inverted mask when returning to the “drawn & parametric mask” mode).

Note: although I don't think it is safe to remove it in the current state of the module, I agree that the blending GUI can be quite confusing at times...

TurboGit commented:
> Based on some testing I've found that blurring and feathering are actually fighting each other and it is not easy to have a properly delimited mask that is smooth inside. Nevertheless, I've made the new order (feather after blur) the default as in most cases it is better to avoid the halos.

Based on my testing the "after blur" is indeed a bit better but using the "output image after blur" seems always a better default than the "input image after blur". Can you double check that ?

TurboGit commented:
Also I would remove image from all choices as image is implied to me. This will makes the combo a bit less wide (especially for the french translation):
- output before blur
- input before blur
- output after blur
- output before blur

Is that ok with you ?

TurboGit commented:
> Now, if we remove the “invert mask” combo, depending on whether we set the “combine masks” combo to “exclusive” or “exclusive & inverted” when using the “drawn & parametric mask” mode, the display of the polarity for the shapes (in “drawn mask” mode) will mean something different.

I was suggesting to remove the combo when "drawn mask" **only** was selected. It must be kept indeed when using drawn + parametric.

hlecleme commented:
Removed “image” from the feathering guide choices. I also updated the label to explain what before/after blur means.

Double checked the code: it selects the right image for feathering. It is possible that using the output image as feathering guide provides better results, but I wouldn't make it a default as it is less “safe”: the mask would then depend on the output of the module, meaning that when changing a parameter of the module both the effect of the module and the pixels where it is applied could change simultaneously.

TurboGit commented:
Something else worth simplifying.

IINM there is duplicate entries in the fusion mode combo. For example: `subtract` and `subtract Inverse`  are redundant as the later can be done with `subtract` and inverting the fusion with the new button.

So I would remove `subtract inverse`, `divide inverse`, and do the migration of old edits in `dt_develop_blend_legacy_params()` if needed.

hlecleme commented:
I would have liked to do that, but unfortunately a reversed "subtract" blending and a subtract inverse blending (the “subtract reverse” blending of darktable 3.4) are not completely equivalent:

```
- subtract:          a * (1.0f - opacity) + (a - p * b) * opacity
- subtract inverse:  a * (1.0f - opacity) + (b - p * a) * opacity
- reversed subtract: b * (1.0f - opacity) + (b - p * a) * opacity
```

This means that the region inside the mask (considering an opacity of 1.0f) would be equivalent between subtract inverse and a reversed subtract, but not the region outside of the mask. It is similar for the divide inverse. That's the reason why I renamed them to “inverse” instead of “reverse”.

We could also hide them (make them deprecated), but I'm not sure to have enough distance (and background) to give an opinion on that.

TurboGit commented:
> We could also hide them (make them deprecated), but I'm not sure to have enough distance (and background) to give an opinion on that.

My first thinking is yes do that. Why ? Because I think no one will be able to understand the differences between the different modes here. I already found that using a specific blend mode is somewhat empirical, but there... well... we are completely in the dark (at least me).

Another way to put this: will somebody say too bad this mode was deprecated as it has nice property to achieve this which is not possible with the new mode + selecting reverse button.

Thoughts?

TurboGit commented:
Now, we may of course delay this deprecation in another PR and possibly after 3.6 even if I'd like to sort out before 3.6.

hlecleme commented:
Or deprecate them now and if someone complains with a valid use case, show them again?

Those blending modes need to stay for compatibility with older edits, but it is quite easy to hide or show them and this may happen without the need of a new `dt_develop_blend_params_t` version.

Technically, it won't be possible to have the reversed subtract or reversed divide effect masked in a “regular” way, i.e. by not affecting the *input* image outside of the masked region.

Shall we go for the deprecation now?

TurboGit commented:
> Shall we go for the deprecation now?

Yes please!

hlecleme commented:
> > Shall we go for the deprecation now?
> 
> Yes please!

Done!

TurboGit commented:
Thanks for the quick turnaround... I'll check that tomorrow and will commit if not more issues/questions on my side.

hlecleme commented:
Thanks for your review! :)

Note: for the “reverse” button (only item 2 in the description), if we want to avoid additional complexity for now (for something that will probably not often be used), we could also limit the change to deprecate the three 3.4 “reverse” blending modes (multiply, subtract and divide) and decide later if it needs to be added. I made this implementation proposal to gather some feedback about its usefulness... @aurelienpierre: have you already used the divide reverse blending mode in your edits?

TurboGit commented:
I think this is ready to be merged. It looks ok to me now for 3.6. @aurelienpierre , no objections ?

TurboGit reviewed (approved):
All good to me. Ready for merging.

aurelienpierre commented:
>  @aurelienpierre: have you already used the divide reverse blending mode in your edits?

Never. I hoped the most advanced users of dt would find some use case for it, but I haven't heard about it so far.

hlecleme commented:
I wonder if it wouldn't be best for darktable 3.6 to remove the “reverse blending” from this PR and to deprecate the “multiply reverse” blending mode (to avoid the confusion as described in #7464), but I would keep the current “subtract reverse” and “divide reverse” blending modes for darktable 3.6.

This would thus avoid the confusion on what image gets blended on top of the other and doesn't complexify the GUI for this release for some feature that probably not many of us will use...

We could then keep this in mind when working on the reorganization of the blending GUI.

elstoc commented:
Someone's going to have to help me with the documentation for all this because I'm kinda lost, reading the discussion :)

TurboGit commented:
> I wonder if it wouldn't be best for darktable 3.6 to remove the “reverse blending” from this PR and to deprecate the “multiply reverse” blending mode (to avoid the confusion as described in #7464), but I would keep the current “subtract reverse” and “divide reverse” blending modes for darktable 3.6.

Not sure. The new proposed UI is cleaner. All modes can be reversed and there is no more specific reverse mode. So I see that like a generalization and find it cleaner. My only concern is about the level of controls on the blending module but we would work that out after 3.6 I would say.

So @hlecleme if you agree I'll merge now as it is on this PR.

hlecleme commented:
Ok for me, thanks!

@elstoc I'll try to take some time next week to propose a PR for the documentation.

TurboGit commented:
Thanks !
